src/aspire/classification/class_selection.py

```diff
@@ -107,11 +146,507 @@ def __init__(self, seed=None):
+        :param n_img: number of images available
+        """
+        return super()._check_selection(selection, n_img, len_operator=le)
+
+
+class NeighborVarianceWithRepulsionClassSelector(
+    GreedyClassRepulsion, NeighborVarianceClassSelector
+):
+    """
+    Selects top classes based on highest contrast with GreedyClassRepulsion.
+    """
+
+
+class GlobalWithRepulsionClassSelector(GreedyClassRepulsion, GlobalClassSelector):
+    """
+    Extends ClassSelector for methods that require
+    passing over all class average images and also GreedyClassRepulsion.
+    """
+
+
+class ImageQualityFunction(ABC):
+    """
+    A callable image quality scoring function.
+
+    The main advantage to using this class is to gain access to a grid
+    caching and Image/Numpy conversion.
+    """
+
+    _grid_cache = {}
+
+    @abstractmethod
+    def _function(self, img):
+        """
+        User defined 1d radial weight function.  The function is
+        expected to be defined on [0,sqrt(2)].  Function range is
+        currently not limited, but [0,1] is favorable.
+
+        Developers can use the self._grid_cache for access to a
+        grid_2d instance matching resolution of img.
+
+        :param img: 2d Numpy array
+        :returns: Image quality score
+        """
+
+    def __call__(self, img):
+        """
+        Given an image instance or a 2D np array,
+        calls the instance's weight function.
+
+        :param img: `Image`, 2d Numpy array,
+            or 3d array where slow axis is stack axis.
+            When called with an image stack>1, an array is returned.
+        :returns: Image quality score(s)
+        """
+
+        if isinstance(img, Image):
+            img = img.asnumpy()
+
+        # Allow the function to be used on a single image presented as
+        # a 2d Numpy array, as well as a stack.
+        if img.ndim == 2:
+            stack_len = 0  # singleton
+            img = img[np.newaxis, :, :]
+        else:
+            stack_len = img.shape[0]
+
+        # Image size.
+        L = img.shape[-1]
+
+        # Generate grid on first call as needed.
+        self._grid_cache.setdefault(L, grid_2d(L, dtype=img.dtype))
+
+        # Call the function over img stack.
+        res = np.fromiter(map(self._function, img), dtype=img.dtype)
+
+        # Return singleton when given a singleton (2d array).
+        if stack_len == 0:
+            res = res[0]
+
+        return res
+
+
+class ContrastImageQualityFunction(ImageQualityFunction):
+    """
+    Computes the variance of pixels.
+    """
+
+    def _function(self, img):
+        """
+        Scoring function based on variance.
+
+        :param img: Input image as 2d Numpy array.
+
+        :return: Pixel variance.
+        """
+        return np.var(img)
+
+
+class BandedSNRImageQualityFunction(ImageQualityFunction):
+    """
+    Computes the ratio of variance of central pixels of image to
+    pixels in a configurable outer band.
```

Comment:
Junk averages tend to either be noisy blobs or have lots of really intense activity near the support boundary.   Using regular contrast also ranks intense boundary junk highly; it has no spatial awareness.  Using regular SNR via our noise estimator doesn't work well because our rotating/averaging leaves the "ears" artificially clean (junk things still seem strong).

So I sort class averages by those that have a strong central signal relative to the signal near the compact support line.  For well behaved averages we are similar to sorting by var... but for boundary junk or noise the ratio is small.

Example. From spending way too much time looking at trash averages of experimental data, many look like below (high contrast class from 10073).
<img width="474" alt="Screen Shot 2023-03-30 at 12 53 35 PM" src="https://user-images.githubusercontent.com/47759732/228909221-35479415-80d2-431f-b3ed-b07b8c7d8dd7.png">

FWIW, I think we can do a similar thing using one of the weighting functions (ramp/bump) with variance.  I think some smart bandpass filtering would be better, but didn't get there yet.

Reply:
Ok got it. So it's essentially replacing the ears with the edge of the disk. Makes sense.